### PR TITLE
update kvdb & co

### PR DESCRIPTION
**Pull request by ordian**
Companion for https://github.com/paritytech/substrate/pull/12312.

We we bail early on IO errors in iterators.

### CODE REVIEW

File: node/subsystem-util/src/database.rs [hunk 1]
```diff
@@ -179,58 +175,47 @@ pub mod paritydb_impl {
...
 		fn iter_with_prefix<'a>(
 			&'a self,
 			col: u32,
 			prefix: &'a [u8],
-		) -> Box<dyn Iterator<Item = (Box<[u8]>, Box<[u8]>)> + 'a> {
+		) -> Box<dyn Iterator<Item = Result<DBKeyValue>> + 'a> {
 			if prefix.len() == 0 {
 				return self.iter(col)
 			}
-			let mut iter = handle_err(self.db.iter(col as u8));
```
Comment (ordian) [inline node/subsystem-util/src/database.rs:205]: @cheme I haven't removed handle_err in `write`, but it would be nice to do eventually.

Comment (cheme) [reply node/subsystem-util/src/database.rs:205]: Oh yes, I missed it in the kvdb PR, next time.

File: node/core/av-store/src/lib.rs [hunk 2]
```diff
@@ -819,10 +820,10 @@ async fn process_block_finalized<Context>(
 			let mut iter = subsystem
 				.db
 				.iter_with_prefix(subsystem.config.col_meta, &start_prefix)
-				.take_while(|(k, _)| &k[..] < &end_prefix[..])
+				.take_while(|r| r.as_ref().map_or(true, |(k, _v)| &k[..] < &end_prefix[..]))
```
[the remark this answers is not in the document]
Comment (ordian) [reply node/core/av-store/src/lib.rs:823]: we will bail on the first error (see `?`)
take_while is not an eager loop, but lazy

File: node/subsystem-util/src/database.rs (hunk 1 above)
Comment (ordian) [reply node/subsystem-util/src/database.rs:205]: I meant to say that part of #4939 is now possible/easier to implement. But I'm leaving it out of scope of this PR.

**Review by cheme — APPROVED** [review]
LGTM

File: node/subsystem-util/src/database.rs [hunk 3]
```diff
@@ -179,58 +175,47 @@ pub mod paritydb_impl {
...
 		}
 
-		fn iter<'a>(&'a self, col: u32) -> Box<dyn Iterator<Item = (Box<[u8]>, Box<[u8]>)> + 'a> {
-			let mut iter = handle_err(self.db.iter(col as u8));
+		fn iter<'a>(&'a self, col: u32) -> Box<dyn Iterator<Item = Result<DBKeyValue>> + 'a> {
+			let mut iter = match self.db.iter(col as u8) {
+				Ok(iter) => iter,
+				Err(e) => return Box::new(std::iter::once(map_err(Err(e)))),
+			};
 			Box::new(std::iter::from_fn(move || {
```
Comment (bkchr) [inline node/subsystem-util/src/database.rs:190]:
```suggestion
			Box::new(iter.map(|r| map_err(r.map(|(k, v)| (k.into(), v))))
```

That should do the same and is easier to read?

File: node/subsystem-util/src/database.rs [hunk 4]
```diff
@@ -179,58 +175,47 @@ pub mod paritydb_impl {
...
-				if let Some((key, value)) = handle_err(iter.next()) {
-					key.starts_with(prefix)
-						.then(|| (key.into_boxed_slice(), value.into_boxed_slice()))
-				} else {
-					None
-				}
+				iter.next().transpose().and_then(|r| {
+					map_err(r.map(|(k, v)| k.starts_with(prefix).then(|| (k.into(), v))))
+						.transpose()
+				})
```
Comment (bkchr) [inline node/subsystem-util/src/database.rs:214]: Similar can be done here.

File: node/core/chain-selection/src/db_backend/v1.rs [hunk 5]
```diff
@@ -254,10 +259,13 @@ impl Backend for DbBackend {
 			self.inner.iter_with_prefix(self.config.col_data, &BLOCK_HEIGHT_PREFIX[..]);
 
 		let val = blocks_at_height_iter
-			.filter_map(|(k, _)| decode_block_height_key(&k[..]))
+			.filter_map(|r| match r {
+				Ok((k, _)) => decode_block_height_key(&k[..]).map(Result::Ok),
```
Comment (bkchr) [inline node/core/chain-selection/src/db_backend/v1.rs:263]:
```suggestion
				Ok((k, _)) => decode_block_height_key(&k[..]).map(Ok),
```

**Review by bkchr — APPROVED** [review]

File: node/subsystem-util/src/database.rs (hunk 3 above)
Comment (ordian) [reply node/subsystem-util/src/database.rs:190]: This doesn't work as parity-db impl doesn't implement iterator, but only [`next`[(https://github.com/paritytech/parity-db/blob/43b451edfdb85e463ab181dd0951872c29b863ac/src/btree/iter.rs#L304) method.
@cheme correct me if I'm wrong

Comment (bkchr) [reply node/subsystem-util/src/database.rs:190]: Ahh okay...

**Comment by ordian** [comment]
bot merge

**Comment by ordian** [comment]
bot merge

**Pull request merged**